### PR TITLE
WIP: Glib take

### PR DESCRIPTION
**Pull request by MJacred**
using glib.Take() and runtime.SetFinalizer() in more places

review is welcome

### CODE REVIEW

File: gtk/gtk_export.go [hunk 1]
```diff
@@ -141,7 +142,9 @@ func goTreeModelForeachFunc(model *C.GtkTreeModel, path *C.GtkTreePath, iter *C.
 	treeModelForeachFuncRegistry.Unlock()
 
 	goPath := &TreePath{(*C.GtkTreePath)(path)}
+	runtime.SetFinalizer(goPath, (*TreePath).free)
 	goIter := &TreeIter{(C.GtkTreeIter)(*iter)}
+	runtime.SetFinalizer(goIter, (*TreeIter).free)
```
Comment (founderio) [inline gtk/gtk_export.go]: Iter should not be freed, see the paragraph "The lifecycle of an iterator can be a little confusing at first..." here: https://developer.gnome.org/gtk3/stable/GtkTreeModel.html#GtkTreeModel.description

The few examples of this function I have found also don't free the treepath, and I'd tend to agree on that - we would litter finalizers for ALL treepaths in that model. Not 100% sure on this, but I assume the caller manages this reference.

Code that needs to retain that path should probably [copy it](https://developer.gnome.org/gtk3/stable/GtkTreeModel.html#gtk-tree-path-copy) which definitely should set a finalizer.

This is probably similar for `goTreeSelectionForeachFunc`

File: gtk/gtk.go [hunk 2]
```diff
@@ -7677,6 +7677,7 @@ func TargetEntryNew(target string, flags TargetFlags, info uint) (*TargetEntry,
 		return nil, nilPtrErr
 	}
 	t := (*TargetEntry)(unsafe.Pointer(c))
+	// FIXME
 	// causes setFinilizer error
```
Comment (founderio) [inline gtk/gtk.go]: This is likely down to the fact that the `TargetEntry` type just points to `C.GtkTargetEntry` and thus is not managed in Go heap.

Above struct should be:
```go
type TargetEntry struct {
    C.GtkTargetEntry
}
```

Then, SetFinalizer will work.

File: glib/settings_schema_source.go [hunk 3]
```diff
@@ -50,15 +53,33 @@ func SettingsSchemaSourceNewFromDirectory(dir string, parent *SettingsSchemaSour
...
 
-	return wrapSettingsSchema(C.g_settings_schema_source_lookup(v.native(), cstr, gbool(recursive)))
+	cSchema := C.g_settings_schema_source_lookup(v.native(), cstr, gbool(recursive))
+	if cSchema == nil {
+		return nil
+	}
+
+	//The finalizer is on the glib.SettingsSchema.
+	s := wrapSettingsSchema(cSchema)
+	runtime.SetFinalizer(s, (*SettingsSchema).Unref)
```
Comment (founderio) [inline glib/settings_schema_source.go]: Who calls `Ref()` here? simply adding `Unref()` may be bad.

File: glib/settings_schema_source.go [hunk 4]
```diff
@@ -50,15 +53,33 @@ func SettingsSchemaSourceNewFromDirectory(dir string, parent *SettingsSchemaSour
...
 
-	return wrapSettingsSchemaSource(C.g_settings_schema_source_new_from_directory(cstr, parent.native(), gbool(trusted), nil))
+	schemaSource := C.g_settings_schema_source_new_from_directory(cstr, parent.native(), gbool(trusted), nil)
+	if schemaSource == nil {
+		return nil
+	}
+
+	//The finalizer is on the glib.SettingsSchemaSource.
+	s := wrapSettingsSchemaSource(schemaSource)
+	runtime.SetFinalizer(s, (*SettingsSchemaSource).Unref)
```
Comment (founderio) [inline glib/settings_schema_source.go]: Who calls `Ref()` here? simply adding `Unref()` may be bad.

File: glib/glib.go [hunk 5]
```diff
@@ -1278,7 +1280,7 @@ func marshalPointer(p uintptr) (interface{}, error) {
 
 func marshalObject(p uintptr) (interface{}, error) {
 	c := C.g_value_get_object((*C.GValue)(unsafe.Pointer(p)))
-	return newObject((*C.GObject)(c)), nil
+	return NewObject((*C.GObject)(c)), nil
```
Comment (founderio) [inline glib/glib.go]: This should probably also call `Take()`?

File: gdk/gdk_since_3_8.go [hunk 6]
```diff
@@ -179,8 +180,11 @@ func (v *FrameTimings) Native() uintptr {
 }
 
 func wrapFrameTimings(ptr unsafe.Pointer) (*FrameTimings) {
-	obj := &glib.Object{glib.ToGObject(ptr)}
-	return &FrameTimings{obj}
+	obj := glib.NewObject(glib.ToGObject(ptr))
+	f := &FrameTimings{obj}
+	// The finalizer is not on the glib.Object but on the FrameTimings
+	runtime.SetFinalizer(f, (*FrameTimings).Unref)
```
Comment (founderio) [inline gdk/gdk_since_3_8.go]: Then who called `Ref()`?

File: gdk/gdk.go [hunk 7]
```diff
@@ -2334,23 +2292,20 @@ func (v *Window) PixbufGetFromWindow(x, y, w, h int) (*Pixbuf, error) {
 		return nil, nilPtrErr
 	}
 
-	obj := &glib.Object{glib.ToGObject(unsafe.Pointer(c))}
-	p := &Pixbuf{obj}
-	//obj.Ref()
-	runtime.SetFinalizer(p, func(_ interface{}) { obj.Unref() })
-	return p, nil
+	obj := glib.Take(unsafe.Pointer(c))
```
Comment (founderio) [inline gdk/gdk.go]: `gdk_pixbuf_get_from_window` is marked as "transfer full". I do think that means we don't call `Ref()` but just ensure to clean up in a finalizer.

File: gdk/gdk.go [hunk 8]
```diff
@@ -1801,10 +1774,8 @@ func PixbufAnimationNewFromFile(filename string) (*PixbufAnimation, error) {
 		return nil, errors.New(C.GoString((*C.char)(err.message)))
 	}
 
-	obj := &glib.Object{glib.ToGObject(unsafe.Pointer(c))}
-	p := &PixbufAnimation{obj}
-	runtime.SetFinalizer(p, func(_ interface{}) { obj.Unref() })
-	return p, nil
+	obj := glib.Take(unsafe.Pointer(c))
```
Comment (founderio) [inline gdk/gdk.go]: [`gdk_pixbuf_animation_new_from_file`](https://developer.gnome.org/gdk-pixbuf/stable/gdk-pixbuf-Animations.html#gdk-pixbuf-animation-new-from-file) returns "A newly-created animation with a reference count of 1, or NULL", which means we do not call `Ref()`, just ensure that `Unref()` is called.

File: gdk/gdk.go [hunk 9]
```diff
@@ -1704,11 +1680,8 @@ func (v *Pixbuf) Flip(horizontal bool) (*Pixbuf, error) {
 		return nil, nilPtrErr
 	}
 
-	obj := &glib.Object{glib.ToGObject(unsafe.Pointer(c))}
-	p := &Pixbuf{obj}
-	//obj.Ref()
-	runtime.SetFinalizer(p, func(_ interface{}) { obj.Unref() })
-	return p, nil
+	obj := glib.Take(unsafe.Pointer(c))
```
Comment (founderio) [inline gdk/gdk.go]: gdk_pixbuf_flip is "transfer full", no calling `Ref()`, just ensure `Unref()` is called.

**Comment by founderio** [comment]
I've left some comments, but have not checked ALL the functions - there are more which I did not add a comment to.

In general, check the documentation of GTK/glib for the ownership handling and reference count.
Generally check these bits here:

image

Hover to see a (brief) explanation. Sometimes these are missing, but the docs mention "reference count":

image

My take on understanding these:
* `transfer none`: Ownership remains with whoever gave you that reference. To retain it, you have to call `Ref()` and later `Unref()` when you don't need it. This is what `glib.Take()` does.
* `transfer full`: Ownership is transferred to you, meaning you have to clean up. `Ref()` will have been called ONCE for you, so don't call it again but ensure to call `Unref()`.  
In our case, this will most probably happen in a finalizer - but if the reference does not escape the wrapper layer, we can call it in wrapper code already, clearing memory earlier.
* Reference count of 1: See `transfer full`.

In all cases I think it would be a good practise to add a comment to the wrapper code indicating which kind of ownership handling is present to avoid confusion in the future.

**Comment by andre-hub** [comment]
@founderio @MJacred 
hello 

what we do with this changes?

**Comment by MJacred** [comment]
I wanted to add some more bindings, so this one got put on hold.

I'll finish this next month.

File: gdk/gdk.go (hunk 7 above)
Comment (MJacred) [reply gdk/gdk.go:2551]: Done

File: gtk/gtk_export.go (hunk 1 above)
Comment (MJacred) [reply gtk/gtk_export.go:146]: Done

File: gdk/gdk.go (hunk 9 above)
Comment (MJacred) [reply gdk/gdk.go:1931]: done

File: gdk/gdk.go (hunk 8 above)
Comment (MJacred) [reply gdk/gdk.go:2025]: done

File: gdk/gdk.go [hunk 10]
```diff
@@ -1907,12 +1875,9 @@ func (v *PixbufLoader) WriteAndReturnPixbuf(data []byte) (*Pixbuf, error) {
 		return nil, nilPtrErr
 	}
 
-	obj := &glib.Object{glib.ToGObject(unsafe.Pointer(c2))}
-	p := &Pixbuf{obj}
+	obj := glib.Take(unsafe.Pointer(c2))
```
[the remark this answers is not in the document]
Comment (MJacred) [reply gdk/gdk.go:2126]: indeed

File: gdk/gdk.go [hunk 11]
```diff
@@ -1940,11 +1905,9 @@ func (v *PixbufLoader) GetPixbuf() (*Pixbuf, error) {
 		return nil, nilPtrErr
 	}
 
-	obj := &glib.Object{glib.ToGObject(unsafe.Pointer(c))}
-	p := &Pixbuf{obj}
 	//obj.Ref() // Don't call Ref here, gdk_pixbuf_loader_get_pixbuf already did that for us.
-	runtime.SetFinalizer(p, func(_ interface{}) { obj.Unref() })
-	return p, nil
+	obj := glib.Take(unsafe.Pointer(c))
```
[the remark this answers is not in the document]
Comment (MJacred) [reply gdk/gdk.go:2157]: OK

File: glib/glib.go (hunk 5 above)
Comment (MJacred) [reply glib/glib.go:1316]: let's go with a glib.Take() for now

File: glib/settings_schema_source.go (hunk 4 above)
Comment (MJacred) [reply glib/settings_schema_source.go:63]: true that. I'll add a call of Ref()

Comment (MJacred) [reply glib/settings_schema_source.go:63]: The func sets the ref count to 1. So we can leave it as-is

https://github.com/GNOME/glib/blob/3499bd7713adcb28b020d36bf06eda984fc26da6/gio/gsettingsschema.c#L311

File: glib/settings_schema_source.go (hunk 3 above)
Comment (MJacred) [reply glib/settings_schema_source.go:80]: has "transfer full" on return -> i.e. don't Ref(), but ensure Unref() via finalizer

File: gtk/gtk.go (hunk 2 above)
Comment (MJacred) [reply gtk/gtk.go:8025]: done

File: gdk/gdk_since_3_8.go (hunk 6 above)
Comment (MJacred) [reply gdk/gdk_since_3_8.go]: removed that

**Comment by MJacred** [comment]
@founderio 

>     * `transfer none`: Ownership remains with whoever gave you that reference. To retain it, you have to call `Ref()` and later `Unref()` when you don't need it. This is what `glib.Take()` does.

If gtk (C) has the ownership, should we Ref()+Unref() (usually via `glib.Take()`) at all?  
some stuff i found:
* https://discourse.gnome.org/t/g-variant-new-and-transfer-none/2472
* https://developer.gnome.org/gobject/stable/annotation-glossary.html
* https://developer.gnome.org/programming-guidelines/stable/memory-management.html.en (did not read this one, yet)

> In all cases I think it would be a good practise to add a comment to the wrapper code indicating which kind of ownership handling is present to avoid confusion in the future.

kept that in mind while cleaning up the commits

File: gdk/gdk.go [hunk 12]
```diff
@@ -857,11 +858,8 @@ func (v *Display) GetMaximalCursorSize() (width, height uint) {
 // GetDefaultGroup is a wrapper around gdk_display_get_default_group().
 func (v *Display) GetDefaultGroup() (*Window, error) {
 	c := C.gdk_display_get_default_group(v.native())
-	if c == nil {
-		return nil, nilPtrErr
-	}
-
-	return &Window{glib.Take(unsafe.Pointer(c))}, nil
+	// transfer none
+	return toWindow(c)
```
Comment (founderio) [inline gdk/gdk.go:862]: You say `transfer none`, but `toWindow` does not `Take()` or `Ref()`.

**Comment by founderio** [comment]
> If gtk (C) has the ownership, should we Ref()+Unref() (usually via glib.Take()) at all?

Not sure on this one. My understanding is that if we don't Ref(), and gtk does Unref(), the object is deleted and our pointer may do unexpected things. So I do think Ref+Unref is the way to go, but we need someone with more knowledge about memory management in glib/gtk.

I left some comments for the latest changes. On top of that there are some compiler issues:
```
glib/glib.go:374:3: too many arguments to return
	have (nil)
	want ()
glib/glib.go:534:2: undefined: c
glib/glib.go:535:41: undefined: c
glib/settings_schema.go:36:1: missing return at end of function
glib/settings_schema_source.go:42:1: missing return at end of function
glib/settings_schema_source.go:60:2: undefined: runtime
glib/settings_schema_source.go:70:9: no new variables on left side of :=
glib/settings_schema_source.go:70:9: cannot use wrapSettingsSchema((func literal)()) (type *SettingsSchema) as type string in assignment
glib/settings_schema_source.go:71:12: invalid operation: schema == nil (mismatched types string and nil)
glib/settings_schema_source.go:76:2: undefined: runtime
```

File: glib/glib.go [hunk 13]
```diff
@@ -529,7 +531,8 @@ func ToGObject(p unsafe.Pointer) *C.GObject {
 
 // Ref is a wrapper around g_object_ref().
 func (v *Object) Ref() {
-	C.g_object_ref(C.gpointer(v.GObject))
+	c = C.g_object_ref(C.gpointer(v.GObject))
+	v = NewObject(ToGObject(unsafe.Pointer(c)))
```
[the remark this answers is not in the document]
Comment (MJacred) [reply glib/glib.go:535]: I see. Done

File: glib/glib.go [hunk 14]
```diff
@@ -368,12 +368,12 @@ func (v *Source) Unref() {
...
 
 // Ref is a wrapper around g_source_ref()
-func (v *Source) Ref() *Source {
+func (v *Source) Ref() {
 	c := C.g_source_ref(v.native())
 	if c == nil {
 		return nil
 	}
-	return (*Source)(c)
+	v = (*Source)(c)
```
[the remark this answers is not in the document]
Comment (MJacred) [reply glib/glib.go:376]: Hm... Indeed it does not work.

someone else did the same. I'll see if I can find and correct that one as well

File: gtk/window.go [hunk 15]
```diff
@@ -71,7 +71,11 @@ func WindowNew(t WindowType) (*Window, error) {
 	if c == nil {
 		return nil, nilPtrErr
 	}
-	return wrapWindow(glib.Take(unsafe.Pointer(c))), nil
+	/*
+	 * "Due to Gtk+ keeping a reference to the window internally, gtk_window_new() does not return a reference to the caller.
+	 * To delete a GtkWindow, call gtk_widget_destroy()."
```
[the remark this answers is not in the document]
Comment (MJacred) [reply gtk/window.go:76]: added

File: glib/settings_schema_source.go [hunk 16]
```diff
@@ -32,12 +32,13 @@ func (v *SettingsSchemaSource) native() *C.GSettingsSchemaSource {
...
 // SettingsSchemaSourceGetDefault is a wrapper around g_settings_schema_source_get_default().
 func SettingsSchemaSourceGetDefault() *SettingsSchemaSource {
+	// transfer none
 	return wrapSettingsSchemaSource(C.g_settings_schema_source_get_default())
 }
 
 // Ref() is a wrapper around g_settings_schema_source_ref().
 func (v *SettingsSchemaSource) Ref() *SettingsSchemaSource {
-	return wrapSettingsSchemaSource(C.g_settings_schema_source_ref(v.native()))
+	v = wrapSettingsSchemaSource(C.g_settings_schema_source_ref(v.native()))
```
[the remark this answers is not in the document]
Comment (MJacred) [reply glib/settings_schema_source.go:41]: done

File: glib/settings_schema.go [hunk 17]
```diff
@@ -32,7 +32,7 @@ func (v *SettingsSchema) native() *C.GSettingsSchema {
 
 // Ref() is a wrapper around g_settings_schema_ref().
 func (v *SettingsSchema) Ref() *SettingsSchema {
-	return wrapSettingsSchema(C.g_settings_schema_ref(v.native()))
+	v = wrapSettingsSchema(C.g_settings_schema_ref(v.native()))
```
[the remark this answers is not in the document]
Comment (MJacred) [reply glib/settings_schema.go:35]: done

File: glib/gvariant.go [hunk 18]
```diff
@@ -102,7 +103,8 @@ func (v *Variant) IsFloating() bool {
 // Reference counting is usually handled in the gotk layer,
 // most applications should not need to call this.
 func (v *Variant) Ref() {
-	C.g_variant_ref(v.native())
+	c := C.g_variant_ref(v.native())
+	v = newVariant((*C.GVariant)(c))
```
[the remark this answers is not in the document]
Comment (MJacred) [reply glib/gvariant.go:107]: done

**Comment by MJacred** [comment]
@founderio: I still can't wrap my head around the ownership. Only read a part of the stuff below

found some resources on memory management / reference counting
* https://developer.gnome.org/programming-guidelines/stable/memory-management.html.en
* https://wiki.gnome.org/Projects/GObjectIntrospection/Annotations#Memory_and_lifecycle_management
* https://developer.gnome.org/gobject/stable/chapter-gobject.html
* https://developer.gnome.org/gobject/stable/gobject-The-Base-Object-Type.html
* https://developer.gnome.org/gtk3/stable/gtk-question-index.html
  * 1.4 + 1.5
* http://www.cs.hunter.cuny.edu/~sweiss/course_materials/csci493.70/lecture_notes/GTK_memory_mngmt.pdf

**Comment by diamondburned** [comment]
> If gtk (C) has the ownership, should we Ref()+Unref() (usually via glib.Take()) at all?

From what I know, yes and no. You likely need to wrap it around something that does reference counting on that then, since Go code can reference it even if it's not supposed to and that code may use it after Gtk has destroyed it.

**Comment by diamondburned** [comment]
I also think this is worth a read: https://github.com/gotk3/gotk3/issues/201#issuecomment-642241588

**Comment by founderio** [comment]
That is an interesting article indeed.

What I gather from it:  
gotk and GJS use fundamentally different wrapping styles:  
GJS has one JS representation for each GObject, at most.  
gotk may very well have multiple wrappers for the same GObject - basically the wrappers are just fancy pointers, the sum of them is basically the reference count.

AFAIK we don't have integration of the same level GJS uses. The GObjects do not reference the Go wrappers (cgo does not like that..), so we don't have quite the same issue they have.  
The issue at hand is the actual reference counting: When creating a wrapper, in which cases do we increase the ref count, and with which function? The ref/unref is meant to prevent deletion of the GObject as long as the wrapper is alive, so we have to keep the ref count correct to not cause memory leaks.

Technical note: The toggle refs mentioned in that article can be created/removed with these: https://developer.gnome.org/gobject/stable/gobject-The-Base-Object-Type.html#g-object-add-toggle-ref

Side note for other memory-related issues in gotk:  
There are ways we can clean up Go memory when an object gets deleted, see the chapter on weak references here: https://developer.gnome.org/gobject/stable/gobject-memory.html (basically a callback which tells us when an object is deleted).  
This could be really useful to delete callbacks for signals and similar from go memory when not needed anymore. Only works with descendants of GObject though, it seems. That would rule out GList, GVariant and similar.

**Pull request closed without merging**